src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,140 @@
+/**
+ * @file batchnorm_impl.hpp
+ * @author Praveen Ch
+ * @author Manthan-R-Sheth
+ *
+ * Implementation of the Batch Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+
+// In case it is not included.
+#include "batch_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm(
+  const size_t size, const double eps) : size(size), eps(eps)
+{
+  weights.set_size(size + size, 1);
+}
+
+
+template<typename InputDataType, typename OutputDataType>
+void BatchNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+  deterministic = false;
+  gamma.fill(1.0);
+  beta.fill(0.0);
+  stats.reset();
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BatchNorm<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  output.reshape(input.n_rows, input.n_cols);
+
+  // Mean and variance over the entire training set will be used to compute
+  // the forward pass when deterministic is set to true.
+  if (deterministic)
+  {
+    mean = stats.mean();
+    variance = stats.var(1);
+  }
+  else
+  {
+    mean = arma::mean(input, 1);
+    variance = arma::var(input, 1, 1);
+
+    for (size_t i = 0; i < output.n_cols; i++)
+    {
+      stats(input.col(i));
+    }
+  }
+
+  for (size_t i = 0; i < output.n_rows; i++)
+  {
+    output.row(i) = arma::as_scalar(gamma.row(i))*((input.row(i)
+                    - arma::as_scalar(mean.row(i)))
+                    / (arma::as_scalar(arma::sqrt(variance.row(i) + eps))))
+                    + arma::as_scalar(beta.row(i));
+  }
```

Comment:
Armadillo typically does calculations much faster when you do then columnwise.  Can you figure out a way to do this by looping over columns instead of rows?

Reply:
In this implementation, it is considered that each column is the training sample and so statistics across rows (which correspond to same dimension) are taken. I think transposing input first and finding statistics over columns, thereby looping over columns and transposing the result back should do the trick.

Reply:
so, shall I implement this, if it sounds fine?

Reply:
No, transposing and then performing the calculation will certainly be slower because the cost of the transpose itself will almost certainly outweigh the slowness of the suboptimal access pattern.  I would instead suggest spending a little time to think about how you can perform this computation in batch for all rows at once.  i.e., instead of looping over each column, can you set the entire `output` matrix in a single Armadillo expression?  It should be possible.

Reply:
Sure, I would like to think on this and come up with something nice. 
Thanks.

Reply:
```
variance = variance + eps;
variance = arma::sqrt(variance);
variance = arma::repmat(variance,1,input.n_cols);
output = arma::repmat(gamma,1,input.n_cols)%((input - arma::repmat(mean,1,input.n_cols))%(1/variance)) + arma::repmat(beta,1,input.n_cols);
```

This transforms the input matrix directly to output matrix. This is should do the trick.
Any views @rcurtin